modified models with jpa

Kasgor opened the pull request:
added changes to all models

Anoromi committed:
create basic repository

Kasgor committed:
modified models with jpa

Anoromi commented on src/main/java/com/example/cult_of_tim/cultoftim/models/Book.java:
```diff
@@ -1,16 +1,23 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
     private String title;
```
Варто додати макимальний розмір.

Anoromi commented on src/main/java/com/example/cult_of_tim/cultoftim/models/Category.java:
```diff
@@ -1,10 +1,28 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private String name;
```
Розмір

Anoromi commented on src/main/java/com/example/cult_of_tim/cultoftim/models/Promotion.java:
```diff
@@ -1,27 +1,37 @@
     private LocalDateTime endDate;
-    private List<Long> userIDs;
+
+    @ManyToMany
+    private List<User> users;
```
До чого у Promotion з'єднання до користувачів?

Anoromi committed:
Merge branch 'init-database' into modelsjpa

# Conflicts:
#	src/main/java/com/example/cult_of_tim/cultoftim/models/Author.java

Anoromi committed:
Update table columns

artemlevv committed:
Added repositories

artemlevv committed:
Additional repositories added

artemlevv committed:
Added PromotionDiscountID

artemlevv committed:
Change Category, Promotion, Purchase services

Kasgor committed:
Changed Author services and Book servicess

Anoromi committed:
Merge remote-tracking branch 'origin/changestorepositores' into modelsjpa

HoWurk commented on cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/repositories/UserRepository.java:
```diff
@@ -0,0 +1,21 @@
+
+    User save(User user);
+
+    void deleteById(UUID id);
```
Most of these methods exist in jparepository and don't need to be duplicated:
![image](https://github.com/Anoromi/cult-of-tim/assets/84322415/f678cbcb-7892-4807-be7b-5b37bc02bd7d)
![image](https://github.com/Anoromi/cult-of-tim/assets/84322415/67674819-6ee0-4ef6-8e2d-9f6ee86bcc05)
etc
Other Repository files, same thing

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java:
```diff
@@ -12,106 +13,86 @@
-        return promotionDao.createPromotion(promotion);
+    public Promotion createPromotion(Promotion promotion) {
+        return promotionRepository.save(promotion);
     }
```
We receive not an entity here, need to change to PromotionCreationDTO or similar stuff
https://www.baeldung.com/java-dto-pattern

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java:
```diff
@@ -12,106 +13,86 @@
     public Promotion updatePromotion(Long id, Promotion updatedPromotion) {
         updatedPromotion.setId(id);
-        return promotionDao.updatePromotion(updatedPromotion);
+        return promotionRepository.save(updatedPromotion);
```
same as create, need possibly update dto

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/models/Author.java:
```diff
@@ -1,9 +1,24 @@
+    private String lastName;
+
+    @ManyToMany
+    private List<Book> books;
```
Either make author contain books, or books contain authors. One of the two.
Same with categories and books.

Anoromi committed:
Fix services

HoWurk commented on cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/dao/UserDao.java:
```diff
@@ -11,17 +11,17 @@
-    void deleteUserById(Long id);
+    void deleteUserById(UUID id);
 
     Optional<UserToken> getUserTokenFor(UUID token);
```
If we move on from UserDao, we either change it fully with UserDaoMock and UserMockService so it actually builds, or don't change it, but still fix it.

Anoromi committed:
Remove useless functions in BookRepository

Anoromi committed:
Fix PromotionDiscountRepository

Kasgor committed:
fixes of repositories

Commented redundant methods of user, author, category, promotion, purchase. Changed Service of promotion

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java:
```diff
@@ -37,17 +38,29 @@ public List<Promotion> getUserPromotions(Long userID) {
     @Override
     public boolean isGlobal(Long promotionID) {
-        return promotionRepository.existsByIdAndGlobal(promotionID, true);
+        return promotionRepository.existsByIdAndGlobalPromotion(promotionID, true);
```
First, change globalPromotion to global (boolean in Promotion).
Second, make method better in PromotionRepository, read comment there.
Third, insert here. If for some reason you need existsbysomething, don't put it in isGlobalMethod.

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/service/PromotionService.java:
```diff
@@ -15,9 +16,9 @@ public interface PromotionService {
+    Promotion createPromotion(String description, LocalDateTime startDate, LocalDateTime endDate, List<PromotionDiscount> discounts, boolean globalPromotion);
 
-    Promotion updatePromotion(Long id, Promotion updatedPromotion);
+    Promotion updatePromotion(Long id, String description, LocalDateTime startDate, LocalDateTime endDate, List<PromotionDiscount> discounts, boolean globalPromotion);
```
this is just plain wrong, again, https://www.baeldung.com/java-dto-pattern
Or leave it as it was previously, and ask lecturer next time.

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/repositories/PurchaseRepository.java:
```diff
@@ -10,13 +10,13 @@
+    //Purchase save(Purchase purchase);
 
-    void deleteById(Long id);
+    //void deleteById(Long id);
```
why comment and not just delete?

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java:
```diff
@@ -10,15 +10,15 @@
     List<Promotion> findByUserId(Long userId);
 
-    boolean existsByIdAndGlobal(Long promotionId, boolean global);
+    boolean existsByIdAndGlobalPromotion(Long promotionId, boolean global);
```
I don't know why this is needed, if you need to check if item exists, then just do boolean exists, if you need to check if this id has global true,
boolean existsByIdAndGlobalPromotionTrue(Long promotionId);
https://www.baeldung.com/spring-data-derived-queries

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java:
```diff
@@ -37,17 +38,29 @@ public List<Promotion> getUserPromotions(Long userID) {
+        updatedPromotion.setEndDate(endDate);
+        updatedPromotion.setDiscounts(discounts);
+        updatedPromotion.setGlobalPromotion(globalPromotion);
         return promotionRepository.save(updatedPromotion);
```
change it back, or fix it, read in interface comment

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java:
```diff
@@ -0,0 +1,24 @@
+
+    //List<Promotion> findAll();
+
+    List<Promotion> findByUserId(Long userId);
```
still doesn't work, no field userId in promotion. if you want to query, @Query

HoWurk commented on src/main/java/com/example/cult_of_tim/cultoftim/repositories/BookRepository.java:
```diff
@@ -0,0 +1,20 @@
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
+
+    List<Book> findByAuthorId(Long authorId);
+
+    List<Book> findByCategoryId(Long categoryId);
```
still doesnt work, findAllByAuthorsContaining, same with category
no fields authorId, categoryId

Kasgor committed:
Deleted comments, Returned old Promotion and deleted connerction to user

Problem with the book still remainig even though I changed it to findAllByAuthorsContaining (same categories).
Problem: "Expected parameter types: List<Author> " even though
List<Book> findAllByAuthorsContaining(List<Author> authors);

Anoromi committed:
fix repositories

Anoromi committed:
Merge remote-tracking branch 'origin/modelsjpa' into modelsjpa

# Conflicts:
#	src/main/java/com/example/cult_of_tim/cultoftim/repositories/AuthorRepository.java
#	src/main/java/com/example/cult_of_tim/cultoftim/repositories/BookRepository.java
#	src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java
#	src/main/java/com/example/cult_of_tim/cultoftim/service/PromotionService.java
#	src/main/java/com/example/cult_of_tim/cultoftim/service/impl/BookMockService.java
#	src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java

Anoromi committed:
fix merge

HoWurk committed:
cleanup + DI change

OleksandrSeverhin committed:
Added addBookWithDiscountToPromotion method, tests in run() and basic methods for addBookWithDiscountToPromotion

HoWurk committed:
fix promotion discount

HoWurk committed:
create dto layer

HoWurk committed:
fix book converter
